gax/src/main/java/com/google/api/gax/tracing/Tracer.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * Copyright 2018 Google LLC
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google LLC nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.tracing;
+
+import org.threeten.bp.Duration;
+
+public interface Tracer {
+  Scope inScope();
+
+  void operationSucceeded();
+
+  void operationFailed(Throwable error);
+
+  void connectionSelected(int id);
+
+  void startAttempt();
+
+  void attemptSucceeded();
+
+  void retryableFailure(Throwable error, Duration delay);
+
+  void retriesExhausted();
+
+  void permanentFailure(Throwable error);
+
+  void receivedResponse();
+
+  void sentRequest();
+
+  void sentBatchRequest(long elementCount, long requestSize);
+
+  interface Scope extends AutoCloseable {
+    @Override
+    void close();
+  }
+
+  enum Type {
+    Unary(false, false),
```

Comment:
Please strongly consider removing this `Type` enum and pass `countRequests` and `countResponses` directly to `newTracer()` factory method. Or pack `countRequests` in something like `TracerSettings` and pass that as argument, because this is what the `Type` is about here: to determine the settings of tracing (and that is it, at least in this WIP implementation). We already have the concept of "settings" in gax-java, so it will be consistent with the rest of the library. 

Also, `Type` creates a "parallel" list of all "tracable" callables (parralel to the actual tracable callables hierarchy), so adding a new tracable callable will also require adding something in this enum, which is odd. If one wants to add a new tracable callable, except having new callable he will also have to edit the "very core" portion of the library. This also makes tracing package indirectly aware of what will be using it, thus declaring a "dependency" on it. It should be an opposite: the tracable callables should know about it tracer, but tracer should not know about tracable callables).